components/uniset/src/uniset.rs

```diff
@@ -0,0 +1,316 @@
+use std::{
+    char::{from_u32, MAX},
+    convert::TryFrom,
+    ops::RangeBounds,
+    slice::Iter,
+};
+
+use super::USetError;
+use crate::utils::{deconstruct_range, is_valid};
+/// Represents the end code point of the Basic Multilingual Plane range, starting from code point 0 , inclusive
+const BMP_MAX: u32 = 0xFFFF;
+
+/// UnicodeSet membership wrapper
+///
+/// Provides exposure to membership functions and constructors from serialized UnicodeSets
+/// and predefined ranges.
+/// Implements an [inversion list.](https://en.wikipedia.org/wiki/Inversion_list)
+#[derive(Debug, PartialEq)]
+pub struct UnicodeSet {
+    // If we wanted to use an array to keep the memory on the stack, there is an unsafe nightly feature
+    // https://doc.rust-lang.org/nightly/core/array/trait.FixedSizeArray.html
+    // Allows for traits of fixed size arrays
+    inv_list: Vec<u32>,
+}
+
+impl TryFrom<Vec<u32>> for UnicodeSet {
+    type Error = USetError;
+
+    fn try_from(set: Vec<u32>) -> Result<Self, Self::Error> {
+        if is_valid(&set) {
+            Ok(UnicodeSet { inv_list: set })
+        } else {
+            Err(USetError::InvalidSet(set))
+        }
+    }
+}
+
+impl UnicodeSet {
+    /// Returns UnicodeSet spanning entire Unicode range
+    ///
+    /// The range spans from `0x0 -> 0x10FFFF` inclusive
+    pub fn all() -> UnicodeSet {
+        UnicodeSet {
+            inv_list: vec![0, (MAX as u32) + 1],
+        }
+    }
+
+    /// Returns UnicodeSet spanning BMP range
+    ///
+    /// The range spans from `0x0 -> 0xFFFF` inclusive
+    pub fn bmp() -> UnicodeSet {
+        UnicodeSet {
+            inv_list: vec![0, BMP_MAX + 1],
+        }
+    }
+
+    /// Yields an iterator of start and stop points of ranges in the UnicodeSet
+    ///
+    /// Example:
+    ///
+    /// ```
+    /// use icu_unicodeset::UnicodeSet;
+    /// use std::convert::TryFrom;
+    /// let example_list = vec![0, 10, 15, 20];
+    /// let example = UnicodeSet::try_from(example_list).unwrap();
+    /// let mut example_ranges = example.ranges();
+    /// assert_eq!(Some(&0), example_ranges.next());
+    /// assert_eq!(Some(&10), example_ranges.next());
+    /// assert_eq!(Some(&15), example_ranges.next());
+    /// assert_eq!(Some(&20), example_ranges.next());     
+    /// assert_eq!(None, example_ranges.next());
+    /// ```
+    pub fn ranges(&self) -> Iter<u32> {
```

Comment:
Would it make sense to make this a private method? Otherwise we're leaking implementation details; it doesn't seem likely to me that a client of this library would benefit from being able to iterate over the inversion list without further structure (for example, as `Range<u32>`s).

Reply:
I'm okay with making it private, but the compiler for now will throw an "method is never used" warning, but I will need it for L2 Builder API UnicodeSet, so temporarily removed.

agreed and changed, but for now removed from the PR.